### PR TITLE
"Rotational Equivariance in Convolutional Neural Networks (CNNs) for Modelling Heat Plumes of Heat Pumps in Groundwater" - Code

### PR DESCRIPTION
**Pull request by pLm-k**

### CODE REVIEW

File: README.md
Comment (JuliaPelzer) [inline README.md:1]: case "augmentation_n < 0" is unclear - which value do you expect? any below 0 or rather -1?

File: README.md [hunk 1]
```diff
@@ -37,6 +37,30 @@
...
+    - **During Inference:** Rotates data points for inference to align with a chosen direction and rotates the predictions back. (Default: `False`)
+  - `--use_ecnn`: Enables the use of equivariant UNet. (Default: `False`)
+  - `--data_n`: Restricts the dataset to `data_n` data points. The test set remains unaffected. <br>
+    Setting `data_n <= 0` or larger than the number of data points results in no restriction. (Default: `-1`)
+
+- **Arguments for Experimenting (Apply before training or inference):**
+  - `--mask`: Applies a mask to all data. (Default: `False`)
+  - `--rotate_inputs`: Rotates all data by the specified angle. (Default: `0`)
+
+- **Logging with wandb**
```
Comment (JuliaPelzer) [inline README.md]: wandb" -> "weights & biases"

File: README.md [hunk 2]
```diff
@@ -37,6 +37,30 @@
...
+  - `--augmentation_n`: Adds augmented versions of each data point by rotating each data point with `augmentation_n` angles uniformly sampled from (0°, 360°). <br>
+    If `augmentation_n < 0`, it adds versions rotated by 90°, 180°, and 270°. (Default: `0`)
+  - `--rotate_inference`: 
+    - **During Training:** Aligns all data points with a chosen direction. (Default: `False`)
+    - **During Inference:** Rotates data points for inference to align with a chosen direction and rotates the predictions back. (Default: `False`)
+  - `--use_ecnn`: Enables the use of equivariant UNet. (Default: `False`)
+  - `--data_n`: Restricts the dataset to `data_n` data points. The test set remains unaffected. <br>
+    Setting `data_n <= 0` or larger than the number of data points results in no restriction. (Default: `-1`)
+
+- **Arguments for Experimenting (Apply before training or inference):**
```
Comment (JuliaPelzer) [inline README.md:56]: all arguments required to set? then put into a single parse argument that sets all these arguments at once

File: README.md [hunk 3]
```diff
@@ -115,4 +139,6 @@
 - directly after paper submission (Oct. '23): cdc41426184756b9b1870e5c0f52d399bee0fae0
 - after clean up, one month after paper submission (Oct. '23): c8da3da
 - release for students to extend_plumes (Mar. '24): ed884f9fb3b8af9808f7abcfee9a0810e8c0fe03, branch release_24
-- release for students to work on first stage (e.g. rotational equivariance) (Mar. '24): 083bb976dfccc52d1, branch release_24
\ No newline at end of file
+- release for students to work on first stage (e.g. rotational equivariance) (Mar. '24): 083bb976dfccc52d1, branch release_24
+- directly after equivariance thesis (Oct. '24): 5b195d73c7429c164ed381710baf3689e0ef9701
+- after cleanup, two weeks after thesis submission (Oct. '24): e386d7e070b82522749cb76d894822e53f8bcbe9
```
Comment (JuliaPelzer) [inline README.md]: after equivariance thesis

File: paths.yaml
Comment (JuliaPelzer) [inline paths.yaml:1]: this is not required

File: processing/rotation.py [hunk 4]
```diff
@@ -0,0 +1,165 @@
...
+    for i in range(data.shape[0]):
+        data_out[i] = data[i]*mask
+    
+    return data_out
+
+#cut out circular field of data for batch
+def mask_batch(batch : torch.tensor) -> torch.tensor:
+    return torch.stack([mask_tensor(data) for data in batch])
+
+# FUNCTIONS TO USE IN COMBINATION WITH GRADIENT, ARTIFACTS OF EARLIER TINKERING ===> UNSUPPORTED!
```
Comment (JuliaPelzer) [inline processing/rotation.py]: what does this mean? If it's not required anymore, just remove it ;)

File: networks/run.sh
Comment (JuliaPelzer) [inline networks/run.sh:1]: is this file required?

File: unittests/test_rotations.py [hunk 5]
```diff
@@ -0,0 +1,78 @@
...
+    a = [0,1]
+    b = [1,0]
+
+    # Expected result
+    exp_angle = 270
+    # Actual result
+    calc_angle = rt.get_rotation_angle(a, b)
+
+    # Test
+    assert calc_angle == exp_angle, "Problem detected in finding angle < 180"
```
Comment (JuliaPelzer) [inline unittests/test_rotations.py]: > 180?

**Review by JuliaPelzer — CHANGES_REQUESTED** [review]
few small remarks to address

File: README.md
Comment (pLm-k) [reply README.md:1]: any below 0 is allowed

File: README.md (hunk 2 above)
Comment (pLm-k) [reply README.md:56]: new argument --equivariance case absorbs --rotate_inference and --use_ecnn, --augmentation_n is automatically set to 0 in case oriented_boxes or ecnn is used
